File: usecases_data/usda_quick_stat/pull_usda_qs.py
# This file pulls data from USDA
import json
import logging
import requests
import pandas as pd

logging.basicConfig(format="%(asctime)-15s %(message)s",
                    filename='call_usda.log',
                    level=logging.INFO)

creds = json.load(open('../credentials.json', 'r'))

URL = "http://quickstats.nass.usda.gov/api/"

# Explore possible values from a variable
params = {"param": "class_desc",
          "key": creds['usda_quickstat_key']}
out = requests.get(
    url=URL + 'get_param_values',
    params=params)

# Estimate the count of variables since QuickStat has a cap
year_start = 1950
year_end = 2018
params = {"sector_desc": "CROPS",
          "commodity_desc": "CORN",
          "group_desc": "FIELD CROPS",
          "short_desc": ["CORN, GRAIN - ACRES HARVESTED",
                         "CORN, GRAIN - PRODUCTION, MEASURED IN BU"],
          "domain_desc": "TOTAL",
          "agg_level_desc": "STATE",
          "year": list(range(year_start, year_end)),
          "key": creds['usda_quickstat_key']}
out = requests.get(
    url=URL + 'get_counts',
    params=params)
est_rows = out.json()['count']
if est_rows > 50000:
    print("too many records, estimating {}, "
          "data call will fail".format(est_rows))
else:
    print('Estimated {} rows of data'.format(est_rows))

# Calling the desired output
usda_resp = requests.get(
    url=URL + 'api_GET',
    params=params)

assert usda_resp.status_code == 200

data = pd.DataFrame(usda_resp.json()['data'])

data.to_csv("state_level_{}_{}_corn_yield.csv".format(year_start, year_end))

File: usecases_data/usda_quick_stat/pull_usda_qs.R
# This file pulls data from USDA
library(httr)
library(jsonlite)

URL <- "http://quickstats.nass.usda.gov/api/"
keys = fromJSON('../credentials.json')
apiKey = keys[['usda_quickstat_key']]

params <- list(param="class_desc",
               key=apiKey)
out <- GET(paste(URL, "get_param_values", sep=""),
           query=params)
char_content <- rawToChar(out$content)
data <- fromJSON(char_content)
data[[params[['param']]]][1:10]

# Estimate the count of variables
year_start <- 1950
year_end <- 2018
years <- year_start:year_end
names(years) <- rep('year', length(years))
short_descs <- c("CORN, GRAIN - ACRES HARVESTED",
                 "CORN, GRAIN - PRODUCTION, MEASURED IN BU")
names(short_descs) <- rep('short_desc', length(short_descs))

params <- list(sector_desc="CROPS",
               commodity_desc="CORN",
               group_desc="FIELD CROPS",
               domain_desc="TOTAL",
               agg_level_desc="STATE",
               key=apiKey)
params <- c(years, short_descs, params)
out = GET(paste(URL, 'get_counts', sep=""),
          accept_json(),
          query=params)
est_rows = content(out)[['count']]
if(est_rows > 50000){
    print(paste("too many records, estimating ", est_rows, 
                " data call will fail"))
} else {
    print(paste('Estimated ', est_rows, ' rows of data'))
}
# Calling the desired output
usda_resp <- GET(url=paste(URL, 'api_GET', sep=""),
                 query=params)

usda_resp$status_code == 200
data <- content(usda_resp)[['data']]

headers <- names(data[[1]])
df <- as.data.frame(do.call(rbind, lapply(data, function(x) unlist(x[headers]))))
constant_vals <- sapply(df, function(x) length(unique(x))) == 1
df_sans_constants <- df[, !constant_vals]
write.csv(df_sans_constants, paste(c("state_level", year_start, year_end, "corn_yield.csv"), collapse="_"), row.names=FALSE)

is_alabama <- df$state_name == "ALABAMA"
is2017 <- df$year == 2017
is_area <- df$short_desc == "CORN, GRAIN - ACRES HARVESTED"
a <- df[is_alabama & is2018 & is_area, ]
dim(a)
